src/OpenTK/Platform/X11/XI2MouseKeyboard.cs

```diff
@@ -167,11 +167,32 @@ internal static bool IsSupported(IntPtr display)
             return false;
         }
 
+        private bool GetIndicatorState(string name)
+        {
+            bool state = false;
+            bool lockState = false;
+            int ndx = 0;
+            IntPtr map = new IntPtr ();
+            bool real = false;
+        
+            IntPtr atom = Functions.XInternAtom(API.DefaultDisplay, name, true);
+            Xkb.GetNamedIndicator(API.DefaultDisplay, atom, out ndx, out state, out map, out real);
+            
+            return state;
+        }
+
         KeyboardState IKeyboardDriver2.GetState()
         {
             lock (Sync)
             {
                 KeyboardState state = new KeyboardState();
+
+                if (Xkb.IsSupported (API.DefaultDisplay))
```

Comment:
For first question, yes, what if some user doesnt have xkb but have xinput2. For second one, I dont know how internally xkb and xinput2 works, I suppouse xkb send me state that all keyboard can influent on(some virtual keyboard state). And of couse X11Keyboard.cs also need this, but it very rare(??) to get xserver without xkb and xinput2 extensions. Also there should be some way to get state without xkb(all linux code based on it), but I don't know how to do it(I tried t ouse xmodmap to figure out what modifier used for what key and Caps work even if you delete it form all mod)